### PR TITLE
Feature/cluster pipeline rebas

### PR DESCRIPTION
**Pull request by vmihailenco**

### CODE REVIEW

File: mpipeline.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
+package redis
+
+// Not thread-safe.
+type MultiPipeline struct {
```
Comment (vmihailenco) [inline mpipeline.go]: Can we rename it to `ClusterPipeline`?

Comment (dim) [inline mpipeline.go]: Sure, go ahead

**Comment by dim** [comment]
I think this one's ready now, I've been using it now for a while and don't have anything to add.

**Comment by vmihailenco** [comment]
@dim I tried to rebase this branch on master, but forgot to pull you latest changes first... As a result your last commits are lost. Can you please push them again?

**Comment by dim** [comment]
Force-pushed again

File: cluster_pipeline.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	cmds   []Cmder
+	client *ClusterClient
+	closed bool
+}
+
+// ClusterPipeline creates a new pipeline which is able to execute commands
+// against multiple shards. Be aware that multi-pipelines are
+// not able to follow redirects and you may receive errors if a node
+// fails or if keys are moved before/during pipeline execution.
+func (c *ClusterClient) ClusterPipeline() *ClusterPipeline {
```
Comment (vmihailenco) [inline cluster_pipeline.go]: Can we rename this to just `ClusterClient.Pipeline`? There is no need to repeat cluster twice...

File: cluster_pipeline.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+	pipes := &clusterPipelines{make(map[string]*Pipeline), c.client}
+
+	// Assign commands to pipelines
+	for _, cmd := range cmds {
+		pipes.process(cmd)
+	}
+	pipes.exec()
+
+	// Detect MOVED errors, try again
```
Comment (vmihailenco) [inline cluster_pipeline.go]: Please add a comment explaining why we don't handle `ASK` response here.

File: cluster_pipeline.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		if err := cmd.Err(); err != nil {
+			return cmds, err
+		}
+	}
+	return cmds, nil
+}
+
+// ------------------------------------------------------------------------
+
+type clusterPipelines struct {
```
Comment (vmihailenco) [inline cluster_pipeline.go]: Can we merge this helper class into `ClusterPipeline`? Its usefulness is doubtful...

**Comment by vmihailenco** [comment]
I think we should handle ask/moved in the `Pipeline`. @dim what do you think?

**Comment by vmihailenco** [comment]
Hm, we can't move ask/moved to the Pipeline, because we don't have access to the slots metadata there.

File: cluster_pipeline.go (hunk 2 above)
Comment (dim) [inline cluster_pipeline.go]: I don't mind, we had a originally a `ClusterClient.Pipeline(key)` which was creating a simple pipeline against a single node, happy to rename.

File: cluster_pipeline.go (hunk 4 above)
Comment (dim) [inline cluster_pipeline.go]: I didn't want to assign this to the pipeline directly as this only lives on Exec and is much easier to test on its own.

**Comment by dim** [comment]
@vmihailenco please check, I'd prefer if you could merge this soon(ish)

**Comment by vmihailenco** [comment]
I have simpler approach to ClusterPipeline, but I am waiting for https://github.com/go-redis/redis/pull/82 to get merged. Sorry that I did not let you know.

**Comment by vmihailenco** [comment]
@dim PTAL.

File: cluster_pipeline.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+func (c *ClusterPipeline) Exec() (cmds []Cmder, retErr error) {
+	if c.closed {
+		return nil, errClosed
+	}
+
+	cmds = c.cmds
+	c.cmds = make([]Cmder, 0, 10)
+	if len(cmds) == 0 {
```
Comment (dim) [inline cluster_pipeline.go]: superminor: you could avoid the (re-)allocation of c.cmds by changing the code to:

```go
if c.closed {
  return nil, errClosed
} else if len(c.cmds) == 0 {
  return []Cmder{}, nil
}
```

File: cluster_pipeline.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+		var addr string
+		if len(addrs) > 0 {
+			addr = addrs[0] // First address is master.
+		}
+
+		cmdsMap[addr] = append(cmdsMap[addr], cmd)
+	}
+
+	for attempt := 0; attempt < c.cluster.opt.getMaxRedirects(); attempt++ {
```
Comment (dim) [inline cluster_pipeline.go]: In this case, it should be `attempt <= c.cluster.opt.getMaxRedirects()`.

File: cluster_pipeline.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+		failedCmds := make(map[string][]Cmder)
+
+		for addr, cmds := range cmdsMap {
+			var client *Client
+			if addr != "" {
+				client = c.cluster.getClient(addr)
```
Comment (dim) [inline cluster_pipeline.go]: Should we make `getClient("")` return `randomClient()`?

Comment (dim) [inline cluster_pipeline.go]: Or at least move that block into a helper method. We are using a very similar logic in `ClusterClient.process`

**Comment by dim** [comment]
That's it, not much to add. Looks good otherwise :)

**Comment by vmihailenco** [comment]
@dim addressed all comments. Merging.

**Pull request merged**